drivers/sensor/iis3dwb/iis3dwb.c

```diff
@@ -0,0 +1,404 @@
+		return fs;
+	}
+
+	if (iis3dwb_set_range_raw(dev, fs) < 0) {
+		LOG_ERR("failed to set accelerometer full-scale");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int iis3dwb_freq_to_odr_val(uint16_t freq)
+{
+	int odr;
+
+	switch (freq) {
+	case 0:
+		odr = 0;
+		break;
+	case 26700:
+		odr = 5;
+		break;
+	default:
+		odr = -EINVAL;
+		break;
+	}
+
+	return odr;
+}
+
+static int iis3dwb_set_odr(const struct device *dev, uint16_t freq)
+{
+	int odr;
+
+	odr = iis3dwb_freq_to_odr_val(freq);
+	if (odr < 0) {
+		return odr;
+	}
+
+	if (iis3dwb_set_odr_raw(dev, odr) < 0) {
+		LOG_ERR("failed to set accelerometer sampling rate");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int iis3dwb_dev_config(const struct device *dev,
+			       enum sensor_channel chan,
+			       enum sensor_attribute attr,
+			       const struct sensor_value *val)
+{
+	switch (attr) {
+	case SENSOR_ATTR_FULL_SCALE:
+		return iis3dwb_set_range(dev, sensor_ms2_to_g(val));
+	case SENSOR_ATTR_SAMPLING_FREQUENCY:
+		return iis3dwb_set_odr(dev, val->val1);
+	default:
+		LOG_DBG("Acc attribute not supported");
+		break;
+	}
+
+	return -ENOTSUP;
+}
+
+static int iis3dwb_attr_set(const struct device *dev,
+			      enum sensor_channel chan,
+			      enum sensor_attribute attr,
+			      const struct sensor_value *val)
+{
+	switch (chan) {
+	case SENSOR_CHAN_ACCEL_X:
+	case SENSOR_CHAN_ACCEL_Y:
+	case SENSOR_CHAN_ACCEL_Z:
+	case SENSOR_CHAN_ACCEL_XYZ:
+		return iis3dwb_dev_config(dev, chan, attr, val);
+	default:
+		LOG_DBG("Attr not supported on %d channel", chan);
+		break;
+	}
+
+	return -ENOTSUP;
+}
+
+static int iis3dwb_sample_fetch(const struct device *dev,
+				 enum sensor_channel chan)
+{
+	struct iis3dwb_data *iis3dwb = dev->data;
+	const struct iis3dwb_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	int16_t buf[3];
+
+	/* fetch raw data sample */
+	if (iis3dwb_acceleration_raw_get(ctx, buf) < 0) {
+		LOG_DBG("Failed to fetch raw data sample");
+		return -EIO;
+	}
+
+	iis3dwb->acc[0] = sys_le16_to_cpu(buf[0]);
+	iis3dwb->acc[1] = sys_le16_to_cpu(buf[1]);
+	iis3dwb->acc[2] = sys_le16_to_cpu(buf[2]);
+
+	return 0;
+}
+
+static const struct sensor_driver_api iis3dwb_driver_api = {
+	.attr_set = iis3dwb_attr_set,
+#if CONFIG_IIS3DWB_TRIGGER
+	.trigger_set = iis3dwb_trigger_set,
+#endif /* CONFIG_IIS3DWB_TRIGGER */
+	.sample_fetch = iis3dwb_sample_fetch,
+	.channel_get = iis3dwb_channel_get,
+};
+
+static int iis3dwb_init(const struct device *dev)
+{
+	struct iis3dwb_data *iis3dwb = dev->data;
+	const struct iis3dwb_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	uint8_t wai;
+	uint8_t filter;
+
+	iis3dwb->dev = dev;
+
+	/* check chip ID */
+	if (iis3dwb_device_id_get(ctx, &wai) < 0) {
+		return -EIO;
+	}
+
+	if (wai != IIS3DWB_ID) {
+		LOG_ERR("Invalid chip ID");
+		return -EINVAL;
+	}
+
+	/* reset device */
+	if (iis3dwb_reset_set(ctx, PROPERTY_ENABLE) < 0) {
+		return -EIO;
+	}
+
+	k_busy_wait(100);
```

Comment:
Can we sleep instead?

Reply:
yes, sure

Reply:
Here I also added the check on bit got reset